server/py/framework/db/sqldb/db.py

```diff
@@ -3194,52 +3200,86 @@ def skip_deletion():
+        total_deleted = 0
 
-        while True:
-            ids_to_delete = (
-                session.query(table.id)
-                .filter(where_clause, table.id > last_id)
-                .order_by(table.id)
-                .limit(batch_size)
-                .all()
-            )
+        logger.info(
+            "starting_batched_range_delete",
+            table=str(table),
+            batch_size=batch_size,
+        )
 
-            if not ids_to_delete:
-                break
+        outer_cm = nullcontext() if session.in_transaction() else session.begin()
 
-            id_values = [row.id for row in ids_to_delete]
+        with outer_cm:
+            while True:
+                start_id = session.scalar(
+                    select(pk).where(where_clause).order_by(pk).limit(1)
+                )
+                if start_id is None:
+                    break
 
-            delete_stmt = (
-                delete(table)
-                .where(table.id.in_(id_values))
-                .execution_options(synchronize_session=False)
-            )
-            result = session.execute(delete_stmt)
-            session.commit()
+                end_id = start_id + batch_size - 1
+                sleep_secs = 0.0
 
-            last_id = id_values[-1]
-            total_deleted += result.rowcount
+                logger.debug("selected_range", start_id=start_id, end_id=end_id)
 
-            logger.debug(
-                "Deleted batch from table",
-                batch_size=len(id_values),
-                total_deleted=total_deleted,
-                last_id=last_id,
-                table=table,
-            )
+                while True:
+                    try:
+                        with session.begin_nested():
+                            result = session.execute(
+                                delete(table)
+                                .where(where_clause)
+                                .where(pk.between(start_id, end_id))
+                                .execution_options(synchronize_session=False)
+                            )
+                            rows = result.rowcount or 0
+                            total_deleted += rows
+
+                        logger.debug(
+                            "range_deleted",
+                            start_id=start_id,
+                            end_id=end_id,
+                            rows_deleted=rows,
+                            total_deleted=total_deleted,
+                        )
+                        break
+
+                    except sqlalchemy.exc.OperationalError as exc:
+                        # MySQL error codes: 1205 = lock-wait timeout, 1213 = deadlock
+                        mysql_code = getattr(exc.orig, "args", [None])[0]
+                        if mysql_code not in (1205, 1213):
+                            raise
+
+                        session.rollback()
+                        sleep_secs = _next_sleep(sleep_secs)
```

Comment:
I think the function under while True can be wrapped within `retry_until_successful` while instead of its `_timeout_exceeded` you can set a custom function for verification - wdyt?